server/src/main/java/org/apache/druid/server/http/CoordinatorCompactionConfigsResource.java

```diff
@@ -119,7 +119,8 @@ public Response addOrUpdateCompactionConfig(
           .getCompactionConfigs()
           .stream()
           .collect(Collectors.toMap(DataSourceCompactionConfig::getDataSource, Function.identity()));
-      newConfigs.put(newConfig.getDataSource(), newConfig);
+      DataSourceCompactionConfig updatedConfig = DataSourceCompactionConfig.from(newConfig, current.getEngine());
+      newConfigs.put(updatedConfig.getDataSource(), updatedConfig);
```

Comment:
Use the `newConfig` instead of the `updatedConfig` to persist.
If the `newConfig` (i.e. the API payload posted by the user) has no engine set, we would have set the engine to the default value in `updatedConfig`.

When we want to switch the engine at the cluster level, we would have to update each config individually.